### PR TITLE
AlignmentsReads

### PR DESCRIPTION
**Pull request by vanetten**
A pipeline for pulling read depth from the Reads API.

### CODE REVIEW

**Comment by pgrosu** [comment]
Hi William/Jason,

Neat practical functionality, but requires a minor fix:

For instance it works for the 1000 Genomes dataset:

```Bash
$ java -cp target/google-genomics-dataflow-*.jar  com.google.cloud.genomics.dataflow.pipelines.AlignmentsReads --project=761052378059  --output=outnew.txt --genomicsSecretsFile=client_secrets.json --shardBAMReading=false  --readGroupSetId=CMvnhpKTFhD04eLE-q2yxnU
$
$ head outnew.txt-00000-of-00001
CMvnhpKTFhD04eLE-q2yxnU 17      41259963        false   60      101
CMvnhpKTFhD04eLE-q2yxnU 17      41273699        true    60      57
CMvnhpKTFhD04eLE-q2yxnU 17      41221835        false   60      101
CMvnhpKTFhD04eLE-q2yxnU 17      41269212        true    60      100
CMvnhpKTFhD04eLE-q2yxnU 17      41234143        true    60      101
CMvnhpKTFhD04eLE-q2yxnU 17      41233039        false   60      101
CMvnhpKTFhD04eLE-q2yxnU 17      41207860        true    60      101
CMvnhpKTFhD04eLE-q2yxnU 17      41259453        true    60      101
CMvnhpKTFhD04eLE-q2yxnU 17      41275846        true    60      101
CMvnhpKTFhD04eLE-q2yxnU 17      41206609        true    60      101
$
```

But does not work for the Illumina Platinum Genomes dataset:

```
$ java -cp target/google-genomics-dataflow-*.jar  com.google.cloud.genomics.dataflow.pipelines.AlignmentsReads  --project=761052378059 --output=outnew2.txt --genomicsSecretsFile=client_secrets.json --shardBAMReading=false  --readGroupSetId=CMvnhpKTFhCAv6TKo6Dglgg
...
INFO: Starting Reads read loop
Exception in thread "main" java.lang.RuntimeException: com.google.cloud.genomics.utils.Paginator$SearchException: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
{
  "code" : 400,
  "errors" : [ {
    "domain" : "global",
    "message" : "The given read group sets are not aligned to any reference with referenceName \"17\". Did you mean \"chr17\"? Wanted one of [\"chrM\" \"chr1\" \"chr2\" \"chr3\" \"chr4\" \"chr5\" \"chr6\" \"chr7\" \"chr8\" \"chr9\" \"chr10\" \"chr11\" \"chr12\" \"chr13\" \"chr14\" \"chr15\" \"chr16\" \"chr17\" \"chr18\" \"chr19\" \"chr20\" \"chr21\" \"chr22\" \"chrX\" \"chrY\" \"*\"]",
    "reason" : "invalidArgument"
  } ],
  "message" : "The given read group sets are not aligned to any reference with referenceName \"17\". Did you mean \"chr17\"? Wanted one of [\"chrM\" \"chr1\" \"chr2\" \"chr3\" \"chr4\" \"chr5\" \"chr6\" \"chr7\" \"chr8\" \"chr9\" \"chr10\" \"chr11\" \"chr12\" \"chr13\" \"chr14\" \"chr15\" \"chr16\" \"chr17\" \"chr18\" \"chr19\" \"chr20\" \"chr21\" \"chr22\" \"chrX\" \"chrY\" \"*\"]"
}
        at com.google.cloud.dataflow.sdk.Pipeline.run(Pipeline.java:151)
        at com.google.cloud.genomics.dataflow.pipelines.AlignmentsReads.main(AlignmentsReads.java:196)
Caused by: com.google.cloud.genomics.utils.Paginator$SearchException: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
{
  "code" : 400,
  "errors" : [ {
    "domain" : "global",
    "message" : "The given read group sets are not aligned to any reference with referenceName \"17\". Did you mean \"chr17\"? Wanted one of [\"chrM\" \"chr1\" \"chr2\" \"chr3\" \"chr4\" \"chr5\" \"chr6\" \"chr7\" \"chr8\" \"chr9\" \"chr10\" \"chr11\" \"chr12\" \"chr13\" \"chr14\" \"chr15\" \"chr16\" \"chr17\" \"chr18\" \"chr19\" \"chr20\" \"chr21\" \"chr22\" \"chrX\" \"chrY\" \"*\"]",
    "reason" : "invalidArgument"
  } ],
  "message" : "The given read group sets are not aligned to any reference with referenceName \"17\". Did you mean \"chr17\"? Wanted one of [\"chrM\" \"chr1\" \"chr2\" \"chr3\" \"chr4\" \"chr5\" \"chr6\" \"chr7\" \"chr8\" \"chr9\" \"chr10\" \"chr11\" \"chr12\" \"chr13\" \"chr14\" \"chr15\" \"chr16\" \"chr17\" \"chr18\" \"chr19\" \"chr20\" \"chr21\" \"chr22\" \"chrX\" \"chrY\" \"*\"]"
}
        at com.google.cloud.genomics.utils.Paginator$4$1$1.apply(Paginator.java:1064)
        at com.google.cloud.genomics.utils.Paginator$4$1$1.apply(Paginator.java:1051)
        at com.google.common.base.Present.transform(Present.java:71)
        at com.google.cloud.genomics.utils.Paginator$4$1.computeNext(Paginator.java:1049)
        at com.google.cloud.genomics.utils.Paginator$4$1.computeNext(Paginator.java:1047)
        at com.google.common.collect.AbstractSequentialIterator.next(AbstractSequentialIterator.java:77)
        at com.google.common.collect.Iterators.advance(Iterators.java:909)
        at com.google.common.collect.Iterables$10.iterator(Iterables.java:865)
        at com.google.common.collect.Iterables$8.iterator(Iterables.java:713)
        at com.google.common.collect.Iterables$8.iterator(Iterables.java:713)
        at com.google.common.collect.Iterables.iterators(Iterables.java:508)
        at com.google.common.collect.Iterables.access$100(Iterables.java:60)
        at com.google.common.collect.Iterables$2.iterator(Iterables.java:498)
        at com.google.cloud.genomics.dataflow.readers.ReadReader.processApiCall(ReadReader.java:58)
        at com.google.cloud.genomics.dataflow.readers.ReadReader.processApiCall(ReadReader.java:29)
        at com.google.cloud.genomics.dataflow.readers.GenomicsApiReader.processElement(GenomicsApiReader.java:58)
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
{
  "code" : 400,
  "errors" : [ {
    "domain" : "global",
    "message" : "The given read group sets are not aligned to any reference with referenceName \"17\". Did you mean \"chr17\"? Wanted one of [\"chrM\" \"chr1\" \"chr2\" \"chr3\" \"chr4\" \"chr5\" \"chr6\" \"chr7\" \"chr8\" \"chr9\" \"chr10\" \"chr11\" \"chr12\" \"chr13\" \"chr14\" \"chr15\" \"chr16\" \"chr17\" \"chr18\" \"chr19\" \"chr20\" \"chr21\" \"chr22\" \"chrX\" \"chrY\" \"*\"]",
    "reason" : "invalidArgument"
  } ],
  "message" : "The given read group sets are not aligned to any reference with referenceName \"17\". Did you mean \"chr17\"? Wanted one of [\"chrM\" \"chr1\" \"chr2\" \"chr3\" \"chr4\" \"chr5\" \"chr6\" \"chr7\" \"chr8\" \"chr9\" \"chr10\" \"chr11\" \"chr12\" \"chr13\" \"chr14\" \"chr15\" \"chr16\" \"chr17\" \"chr18\" \"chr19\" \"chr20\" \"chr21\" \"chr22\" \"chrX\" \"chrY\" \"*\"]"
}
        at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:145)
        at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:113)
        at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:40)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1.interceptResponse(AbstractGoogleClientRequest.java:312)
        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1049)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:410)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:343)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:460)
        at com.google.cloud.genomics.utils.RetryPolicy.execute(RetryPolicy.java:102)
        at com.google.cloud.genomics.utils.Paginator$4$1$1.apply(Paginator.java:1055)
        at com.google.cloud.genomics.utils.Paginator$4$1$1.apply(Paginator.java:1051)
        at com.google.common.base.Present.transform(Present.java:71)
        at com.google.cloud.genomics.utils.Paginator$4$1.computeNext(Paginator.java:1049)
        at com.google.cloud.genomics.utils.Paginator$4$1.computeNext(Paginator.java:1047)
        at com.google.common.collect.AbstractSequentialIterator.next(AbstractSequentialIterator.java:77)
        at com.google.common.collect.Iterators.advance(Iterators.java:909)
        at com.google.common.collect.Iterables$10.iterator(Iterables.java:865)
        at com.google.common.collect.Iterables$8.iterator(Iterables.java:713)
        at com.google.common.collect.Iterables$8.iterator(Iterables.java:713)
        at com.google.common.collect.Iterables.iterators(Iterables.java:508)
        at com.google.common.collect.Iterables.access$100(Iterables.java:60)
        at com.google.common.collect.Iterables$2.iterator(Iterables.java:498)
        at com.google.cloud.genomics.dataflow.readers.ReadReader.processApiCall(ReadReader.java:58)
        at com.google.cloud.genomics.dataflow.readers.ReadReader.processApiCall(ReadReader.java:29)
        at com.google.cloud.genomics.dataflow.readers.GenomicsApiReader.processElement(GenomicsApiReader.java:58)
        at com.google.cloud.dataflow.sdk.util.DoFnRunner.processElement(DoFnRunner.java:132)
        at com.google.cloud.dataflow.sdk.transforms.ParDo.evaluateHelper(ParDo.java:1059)
        at com.google.cloud.dataflow.sdk.transforms.ParDo.evaluateSingleHelper(ParDo.java:963)
        at com.google.cloud.dataflow.sdk.transforms.ParDo.access$000(ParDo.java:441)
        at com.google.cloud.dataflow.sdk.transforms.ParDo$1.evaluate(ParDo.java:951)
        at com.google.cloud.dataflow.sdk.transforms.ParDo$1.evaluate(ParDo.java:946)
        at com.google.cloud.dataflow.sdk.runners.DirectPipelineRunner$Evaluator.visitTransform(DirectPipelineRunner.java:619)
        at com.google.cloud.dataflow.sdk.runners.TransformTreeNode.visit(TransformTreeNode.java:200)
        at com.google.cloud.dataflow.sdk.runners.TransformTreeNode.visit(TransformTreeNode.java:196)
        at com.google.cloud.dataflow.sdk.runners.TransformHierarchy.visit(TransformHierarchy.java:109)
        at com.google.cloud.dataflow.sdk.Pipeline.traverseTopologically(Pipeline.java:204)
        at com.google.cloud.dataflow.sdk.runners.DirectPipelineRunner$Evaluator.run(DirectPipelineRunner.java:592)
        at com.google.cloud.dataflow.sdk.runners.DirectPipelineRunner.run(DirectPipelineRunner.java:328)
        at com.google.cloud.dataflow.sdk.runners.DirectPipelineRunner.run(DirectPipelineRunner.java:70)
        at com.google.cloud.dataflow.sdk.Pipeline.run(Pipeline.java:145)
        at com.google.cloud.genomics.dataflow.pipelines.AlignmentsReads.main(AlignmentsReads.java:196)
$
```

It's not a major fix to perform, since the reference names in the Illumina Platinum Genomes dataset just require a `chr` as a prefix for their `referenceName`.

Hope it helps,
Paul

File: src/main/java/com/google/cloud/genomics/dataflow/pipelines/AlignmentsReads.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+  }
+  
+  public static void main(String[] args) throws GeneralSecurityException, IOException {
+    // Register the options so that they show up via --help
+    PipelineOptionsFactory.register(AlignReadsOptions.class);
+    options = PipelineOptionsFactory.fromArgs(args).withValidation().as(AlignReadsOptions.class);
+    // Option validation is not yet automatic, we make an explicit call here.
+    GenomicsDatasetOptions.Methods.validateOptions(options);
+
+    auth = GCSOptions.Methods.createGCSAuth(options);
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/dataflow/pipelines/AlignmentsReads.java]: The code appears to work better if we switch to:
auth = GenomicsOptions.Methods.getGenomicsAuth(options);

File: src/main/java/com/google/cloud/genomics/dataflow/pipelines/AlignmentsReads.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+import com.google.cloud.genomics.utils.Contig;
+import com.google.cloud.genomics.utils.GenomicsFactory;
+import com.google.cloud.genomics.utils.Paginator;
+import com.google.cloud.genomics.utils.Paginator.ShardBoundary;
+import com.google.common.base.Function;
+import com.google.common.base.Splitter;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+
+import org.codehaus.jackson.annotate.JsonIgnore;
```
Comment (jean-philippe-martin) [inline src/main/java/com/google/cloud/genomics/dataflow/pipelines/AlignmentsReads.java]: This should be:
import com.fasterxml.jackson.annotation.JsonIgnore;

in order for the @JsonIgnore to work in the PipelineOptions.

**Comment by vanetten** [comment]
Woot!

The following is working for me.
What is the correct "references" for an entire chromosome?
chromosome:0:<some really big number?>

Help says that if I want to run on all readGroupSetId(s) that I should set readGroupSetId = '',
but I'm guessing I must need to provide more CLI args to identify the dataset?

PROJECT=idyllic-analyst-574
bucket=gs://mssng_read_depth
OUTPUT=${bucket}/output_align.txt
STAGING=${bucket}/staging
# REFERENCES=MT:0:16569
REFERENCES=22:0:52000000
# --project=${PROJECT}  --output=${OUTPUT}  --genomicsSecretsFile=client_secret.json
#java -cp target/google-genomics-dataflow-*.jar   com.google.cloud.genomics.dataflow.pipelines.AlignmentsReads  --help='com.google.cloud.genomics.dataflow.pipelines.AlignmentsReads$CountReadsOptions'

#--BAMFilePath=gs://bioteamdataflow/test.bam 
# readgroupsetid=CMvnhpKTFhD04eLE-q2yxnU
readgroupsetid=CNOc5OP9ERDzhNPG3_rwmZ4B
java -cp target/google-genomics-dataflow-*.jar \
  com.google.cloud.genomics.dataflow.pipelines.AlignmentsReads \
  --runner=BlockingDataflowPipelineRunner \
  --project=${PROJECT} \
  --stagingLocation=${STAGING} \
  --output=${OUTPUT} \
  --genomicsSecretsFile=client_secrets.json \
  --shardBAMReading=false \
  --readGroupSetId=${readgroupsetid} \
  --references=${REFERENCES} \
  --numWorkers=4

**Comment by pgrosu** [comment]
Good catch, but something seems funny about the scalability of Dataflow with the number of workers.  So I tried the following, and it should have finished fairly quickly, but it was still running after about 15 minutes - which I had to stop:

```
java -cp target/google-genomics-dataflow-*.jar \
com.google.cloud.genomics.dataflow.pipelines.AlignmentsReads \
--project=761052378059 --output=largeoutput.txt \
--genomicsSecretsFile=client_secrets.json --shardBAMReading=false \
--readGroupSetId=CMvnhpKTFhCAv6TKo6Dglgg --references=chr1:1:20000000 \
--numWorkers=100000
```

~p

**Comment by pgrosu** [comment]
Even setting `--shardBAMReading=true` did not seem to help, though I assume this only applies to actual BAM files not `ReadGroupSets`.

**Comment by dionloy** [comment]
You need to use --runner=BlockingDataflowPipelineRunner otherwise you are
just executing on your own machine.

**Comment by pgrosu** [comment]
Thanks Dion for helping me on this.  Do I need Google Cloud Storage for that?  I didn't want to to turn that on and start getting billed after the trial period as we are still troubleshooting this part of the project.  Is there any other way to test this to ensure scalability of the analysis that users might perform on large datasets?  It would be nice to show proof-of-concept use cases.  I'd be happy to help with that.

Thanks,
Paul

**Comment by dionloy** [comment]
Yep everything needs billing =). We have run the example pipelines on
several hundred machines already, so we are good there.

Bill will be withdrawing this PR, as it was a scratch PR for sharing some
work in progress code. Thanks!

**Comment by pgrosu** [comment]
Sounds good, thanks Dion for the confirmation.

Paul

**Pull request closed without merging**